### PR TITLE
refactor(semantic): add binder for FormalParameters and RestElement, replacing the binder for FormalParameters

### PR DESCRIPTION
**Pull request by Dunqing**
Similar with #2013.

This way we will be able to find the corresponding Ast accurately.

The snapshots look a little strange but the change is expected.

### CODE REVIEW

**Comment by Dunqing** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #2114** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/2114?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/2114?utm_source=stack-comment).

File: tasks/coverage/parser_babel.snap [hunk 1]
```diff
@@ -6833,8 +6833,8 @@ Expect to Parse: "typescript/types/const-type-parameters-babel-7/input.ts"
    ╭─[esprima/es2015-arrow-function/complex-rest-in-arrow-not-allowed/input.js:1:1]
  1 │ (a,...[a]) => 0;
    ·  ┬     ┬
-   ·  │     ╰── It can not be redeclared here
-   ·  ╰── `a` has already been declared here
+   ·  │     ╰── `a` has already been declared here
```
Comment (Boshen) [inline tasks/coverage/parser_babel.snap:6836]: Why is the order reversed? Is this easy to fix?

**Comment by Boshen** [comment]
> The snapshots look a little strange but the change is expected.

It's weird for the user.

File: tasks/coverage/parser_babel.snap (hunk 1 above)
Comment (Dunqing) [reply tasks/coverage/parser_babel.snap:6836]: Looking at the visit order, we'll go to `FormalParameters` first followed by `FormalParameter`, which causes the order to change, the fix for this I've thought of so far is to execute the `FormalParameters`' binder method during the `leave_kind` phase, but I'm not sure if that will cause any other problems

**Comment by Boshen** [comment]
### Merge activity

* **Jan 22, 1:23 AM**: @Boshen started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/2114).

**Pull request merged**